### ISSUE TITLE
“Attached” hashtags

### ISSUE THREAD

**Issue by cd24** [mastodon/mastodon#25459]
### Pitch

Mastodon can provide a mechanism to add tags separately from the body of a post. This would likely require updating the format of a post, and is inspired by a format people already follow. There are some posters who include a large variety of hashtags underneath their posts, and it can create a lot of visual clutter in the UI, just to ensure the post reaches the right audience.

For example, this is a common posting format to include a wide variety of hashtags that may be relevant (since you might want to include both ‘#dog’ and ‘#dogs’):

“””
My #puppies really can’t seem to sit still! So glad to see they’re up and #energetic

#dog #dogs #cute #BabyAnimal #BabyAnimals #LOCATION… <ETC>
“””

This style of writing isn’t very natural, and if the hashtag you want doesn’t make grammatical sense you just have to add it to the bottom of the post in a block (especially if there are many variants)

Mastodon could improve this experience by providing a mechanism for any user to “attach” one or more tags to a post separately from the body. There are a few advantages to a system like this: 

1) Others adding hashtags for you. If hashtags can be attached by others, separately from replies, then we can minimize the number of “this should be tagged with #SomeTag” replies to any given post. Attachments would need to be attributed to the person who added the tag, but this could be extremely useful for allowing communities to label content without needing to click through a comment to get to the original post.
2) The ability to add personal tags for discoverability. If you have a series of tags that make sense to you (or a small group of your friends), then you can decorate the post itself with the tags, rather than requiring a full comment. This improves the ergonomics of linking content for yourself and/or a small group via tags.
3) Improved reading ergonomics. Currently, the best way to do either (1) or (2) is via a comment on their post. This works well until the post is inaccessible to a third federated server (showing no content above the reply), or when the comment appears in someone’s timeline (via a followed hashtag), requiring them to click on the reply to see the original post. Attaching tags to the post avoids the first case since there’s no post to display if the original is inaccessible and the second by displaying the tagged post first and foremost.
4) Authors don’t need to clutter their posts with hashtags in order to benefit from hashtag search. This allows authors to write naturally, and add the tagging separately, later, or never.

### Considerations around safety

This can be thought of as a special type of comment, and we should have safety tools to match. At least, I would expect this to have behavior like:

1) If someone is blocked, they cannot attach tags; 
2) If someone is muted, their tag attachments don’t appear - even if the user follows an attached hashtag;
3) The UI clearly labels which account attached the hashtags (whether it’s the author or someone else, and a link to their accounts);
4) If a server is deferederated, their attachments are ignored (excepting users that follow remote users in the limited federation case); and
5) Tag attachments can be removed or blocked by moderators & users (as with replies today)

### Motivation

Finding and using the appropriate hashtag can be difficult, and sometimes you won’t discover it until someone else adds a hashtag as a comment underneath your post. We could benefit from a system where the author and their community can update a “tag” field, and only the tag field, of a post when they discover new hashtags they would like to use.

**Comment by Timothyjchambers** [comment]
Think this is a solid reuest. Frendica and KBIN already do this or things quite like it in terms of Fediverse platforms. Mastodon should too.

**Comment by ThisIsMissEm** [comment]
I'd definitely like to see the Mastodon UI support rendering tags separately from posts. I do also find the idea of allowing users to "suggest" a hashtag for a post an interesting idea.

**Comment by cd24** [comment]
Based on some offline comments, I would like to clarify that these added tags are not modifications to the original post. They are an addition to the post that any user can add (like a comment). These could be rendered as comments to the original post if the software doesn’t support them, but I expect Mastodon would add some more ergonomics to make them easy to use. Perhaps even adding a boost-style line above them like “<ACCOUNT> tagged this post with #TAG” when the post is in the feed because the tag was added, or a footer to the effect of “<ACCOUNT>\n\n#TAG1 #TAG2…etc” if a user follows <ACCOUNT>

**Comment by ThisIsMissEm** [comment]
@cd24 I'd actually suggest that these just generate a suggestion to update your post to include them as a notification, that way you can choose whether you want your post to be on a specific hashtag or not — without another user forcible adding your post to that hashtag. (I'm not sure how we'd represent this in ActivityPub though, maybe Offer or Add activities, but that's a much longer future thing)

**Comment by cd24** [comment]
I think there’s value in allow people to tag content without the author being involved. I think having the author in the middle could get messy, especially since I expect it would be harder to moderate a suggestion than a “post.” I expect this could be a more specialized variant of a quote post, focused only on adding metadata for you, your server, and your followers (but pushing the tags below the content, rather than above it like a typical quote post)…

But I would also be absolutely fine with the Tumblr style of tagging, adding an area for tags at the very bottom of everything rather than in the post body.

**Comment by trwnh** [comment]
isnt this mostly a duplicate of #10743

**Comment by cd24** [comment]
It is very similar, but I think it goes a little further. I think that change + quote posts could be a solution, since an empty quote post with tags as described there could cover all my use cases!

**Comment by nemobis** [comment]
> Fr[i]endica and KBIN already do this

And it causes much confusion when people receive posts in their home timeline without having any way to figure out why. https://github.com/mastodon/mastodon/issues/19528

**Comment by Dehelssey** [comment]
an elegant graphic solution has been proposed here → #25873

**Comment by sam1919** [comment]
Agreed. Also the solution in the previous comment is worth a look.
Please implement that.
